contracts/asset-proxy/contracts/src/ERC20BridgeProxy.sol

```diff
@@ -0,0 +1,128 @@
+import "./interfaces/IAssetProxy.sol";
+import "./interfaces/IERC20Bridge.sol";
+
+
+contract ERC20BridgeProxy is
+    IAssetProxy,
+    Authorizable
+{
+    using LibBytes for bytes;
+    using LibSafeMath for uint256;
+
+    // @dev Result of a successful bridge call.
+    bytes4 constant public BRIDGE_SUCCESS = 0xb5d40d78;
+    // @dev Id of this proxy.
+    //      bytes4(keccak256("ERC20BridgeProxy(address,address,bytes)"))
+    bytes4 constant private PROXY_ID = 0x37708e9b;
+
+    /// @dev Calls a bridge contract to transfer `amount` of ERC20 from `from`
+    ///      to `to`. Asserts that the balance of `to` has increased by `amount`.
+    /// @param assetData Abi-encoded data for this asset proxy encoded as:
+    ///          abi.encodeWithSelector(
+    ///             bytes4 PROXY_ID,
+    ///             address tokenAddress,
+    ///             address bridgeAddress,
+    ///             bytes bridgeData
+    ///          )
+    /// @param from Address to transfer asset from.
+    /// @param to Address to transfer asset to.
+    /// @param amount Amount of asset to transfer.
+    function transferFrom(
+        bytes calldata assetData,
+        address from,
+        address to,
+        uint256 amount
+    )
+        external
+        onlyAuthorized
+    {
+        // Extract asset data fields.
+        (
+            address tokenAddress,
+            address bridgeAddress,
+            bytes memory bridgeData
+        ) = abi.decode(
+            assetData.sliceDestructive(4, assetData.length),
+            (address, address, bytes)
+        );
+
+        // Remember the balance of `to` before calling the bridge.
+        uint256 balanceBefore = balanceOf(tokenAddress, to);
+        // Call the bridge, who should transfer `amount` of `tokenAddress` to
+        // `to`.
+        bytes4 success = IERC20Bridge(bridgeAddress).transfer(
+            bridgeData,
+            tokenAddress,
+            from,
+            to,
+            amount
+        );
+        // Bridge must return the magic bytes to indicate success.
+        require(success == BRIDGE_SUCCESS, "BRIDGE_FAILED");
+        // Ensure that the balance of `to` has increased by at least `amount`.
+        require(
+            balanceBefore.safeAdd(amount) <= balanceOf(tokenAddress, to),
+            "BRIDGE_UNDERPAY"
+        );
+    }
+
+    /// @dev Gets the proxy id associated with this asset proxy.
+    /// @return proxyId The proxy id.
+    function getProxyId()
+        external
+        pure
+        returns (bytes4 proxyId)
+    {
+        return PROXY_ID;
+    }
+
+    /// @dev Retrieves the balance of `owner` for this asset.
+    /// @return balance The balance of the ERC20 token being transferred by this
+    ///         asset proxy.
+    function balanceOf(bytes calldata assetData, address owner)
+        external
+        view
+        returns (uint256 balance)
+    {
+        (address tokenAddress, ,) = abi.decode(
```

Comment:
This can be made more efficient with:

```
(address tokenAddress) = abi.decode(
    assetData.sliceDestructive(4, assetData.length),
    (address)
);
```

Reply:
Yeah, I was trying to sneak in some extra asset data validation this way, but I guess since it's not checking the selector, it's not even doing that very well.